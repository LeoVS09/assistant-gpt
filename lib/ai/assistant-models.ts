/** Output of assistant based on user input */
export interface AssistantOutput {
    /** Response from assitant to the user */
    response: string;
    /** 
     * List of the commands to execute, must be based on user input. 
     * Can be empty 
     * */
    commands?: Array<AbstractCommand>;
}

/** Base abstract interface for commands */
export interface AbstractCommand<Args = object> {
    /** Id of the command to execute */
    id: string;
    /** Arguments which need pass to command execution logic */
    args: Args;
}

/** Comand which ask save command arguments to database */
export interface SaveCommand extends AbstractCommand<SaveCommandArgs> {
    id: "save";

    /** Arguments for save command, will be saved to database as it is */
    args: SaveCommandArgs;
}

/** 
 * Arguments for save command, will be saved to database as it is.
 * Additional data and fields can be added to this interface by assistant if need for user case
 * */
export interface SaveCommandArgs {
    /** Text of the note or task */
    text: string;

    /** 
     * List of the entities which present in text, usefull to query by entity 
     * For example: "Buy milk today" -> ["milk", "today"]
     * */
    entities?: Array<string>;

    /** 
     * List of the tags which can be build based on text, usefull to group multiple notes together 
     * For example: "Buy milk today" -> ["shopping list", "minor priority"]
     * Base tags to group in different lists can be: "shopping list", "reading list", "work list", "personal list"
     * Base tags to group by priority can be: "criticial priority", "high priority", "medium priority", "low priority"
     * Can be extended or adjusted by user
     * Additional tags, specific for user can be added as well
     * */
    tags?: Array<string>;

    /**
     * Status of the note or task, can be "to-do", "in-progress", "done"
     * Can be extended or adjusted by user
     * */
    status?: "string";

    /**
     * Deadline of the note or task. Represented as string in date format
     * */
    deadline?: string;
}


/** Comand which ask execute find query in database with given arguments */
export interface QueryCommand extends AbstractCommand<any> {
    id: "query";

    /** Arguments for query command, will be used to execute find query in database */
    args: any;
}

/** Comand which ask execute update command in database with given arguments */
export interface UpdateCommand extends AbstractCommand<UpdateCommandArgs> {
    id: "update";

    /** 
     * Arguments for update command, will be used to execute update query in database 
     * For example: What I need work on this week? -> {"status":"to-do","deadline":{"$lte":"April 9, 2023"}}
     * */
    args: UpdateCommandArgs;
}

/** Arguments for update command, will be used to execute update query in database */
export interface UpdateCommandArgs {
    /**
     * Query to find document to update
     * For example: I read book about strategies in business this week -> { tags: 'reading list', status: "to-do" }
     * */
    filter: any;

    /**
     * Update query to update document
     * For example: I read book about strategies in business this week -> { "$set": { "status": "done" } }
     * */
    update: any;
}