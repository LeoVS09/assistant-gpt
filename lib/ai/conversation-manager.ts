import { ChatGPTAPI, ChatMessage, SendMessageOptions } from "chatgpt";
import { v4 as uuidv4 } from 'uuid'

/** Message in chat which is sent to AI Agent */
export interface ConversationMessage extends Pick<SendMessageOptions, 'name' | 'onProgress' | 'systemMessage'> {
    text: string
}

/**
 * Interface for AiApi class. Real implementation is in chatgpt/ChatGPTAPI
 */
export interface AiAgent {
    /**
     * Sends a message to the AI agent and receives a response.
     *
     * @param message The message to send.
     * @param options Additional options for sending the message.
     *
     * @returns The response message from the AI agent.
     */
    sendMessage(message: string, options?: SendMessageOptions): Promise<ChatMessage>;
  }
  

/** Manager which maintain persistant conversation with AI Agent */
export class ConversationManager {

    public conversationId = uuidv4()
    public previusMessageId?: string

    constructor(
        private readonly aiAgent: AiAgent,
    ) { 
        // TODO: restore and save conversation into persistent storage
    }

    public async send({text, ...rest}: ConversationMessage): Promise<ChatMessage> {
        const res = await this.aiAgent.sendMessage(text, {
            ...rest,
            conversationId: this.conversationId,
            parentMessageId: this.previusMessageId,
        })

        console.debug(`- User: ${text}\n-Assistant: ${res.text}\n`)

        this.previusMessageId = res.id

        return res;
    }
}