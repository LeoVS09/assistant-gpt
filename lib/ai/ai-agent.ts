import { ChatGPTAPI } from 'chatgpt'
import { AiAgent } from './conversation-manager'


/** Generate AI Agent instance which able perform chat completition */
export const generateAiAgent = (): AiAgent => {
    console.log('env OPENAI_API_KEY is exists', !!process.env.OPENAI_API_KEY)

    return new ChatGPTAPI({
        apiKey: process.env.OPENAI_API_KEY!,
        completionParams: {
            model: "gpt-3.5-turbo",
            temperature: 0.2 // based on https://arxiv.org/pdf/2303.17580.pdf
        }
    
    })
}