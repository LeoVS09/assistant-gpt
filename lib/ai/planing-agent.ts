import { ChatMessage } from "chatgpt"
import { errorToString, generatePromptToFixJsonError, generateSystemPrompt } from "./prompt"
import { retry } from "./retry"
import { DateTime } from "luxon"
import { ConversationManager, ConversationMessage } from "./conversation-manager"
import { AssistantOutput } from "./assistant-models"


export interface CommandsPlan extends AssistantOutput {
    original: ChatMessage
}



export class CommandsParsingError {
    constructor(
        public readonly message: ChatMessage,
        public readonly originalError: any,
    ) { }

    public toString() {
        return `Cannot parse commands, catched error '${errorToString(this.originalError)}'`
    }
}

/** Agent which plan commands based on user input text */
export class PlaningAgent {

    public systemMessage: string
    public appologySeparator = '%RESULT%'

    constructor(
        private readonly chat: ConversationManager,
        private readonly now: DateTime
    ) { 
        this.systemMessage = generateSystemPrompt(this.now)
    }

    public async plan(message: ConversationMessage): Promise<CommandsPlan> {
        return retry(
            async () => {
                // No errors, initial call
                const original = await this.chat.send({
                    ...message,
                    systemMessage: this.systemMessage,
                })
                
                try {
                    // TODO: add validation
                    const { response, commands } = JSON.parse(original.text) as AssistantOutput

                    return { original, response, commands }
                } catch (error) {
                    console.warn('Error during execution', error, 'for response', original.text)
                    throw new CommandsParsingError(original, error)
                }
            },
            async (error) => {
                if (!(error instanceof CommandsParsingError)) {
                    throw error
                }

                // Ouput probably was not valid json
                const retryRes = await this.chat.send({
                    ...message,
                    text: this.genPromptToFixJsonError(error.originalError),
                    systemMessage: this.systemMessage,
                })

                const commandsJson = this.extractCommandsAfterRetry(retryRes.text)
                const { response, commands } = JSON.parse(commandsJson) as AssistantOutput

                return { original: retryRes, response, commands }
            }
        )
    }

    private genPromptToFixJsonError(error: any) {
        return generatePromptToFixJsonError(this.appologySeparator, this.now, error)
    }

    private extractCommandsAfterRetry(text: string) {
        return extractCommandsAfterRetry(this.appologySeparator, text)
    }
}

/** Extract real result from text after given separator */
export const extractCommandsAfterRetry = (separator: string, text: string) => {
    const index = text.lastIndexOf(separator)
    if (index === -1) {
        throw new Error(`Cannot find separator '${separator}' in text '${text}'`)
    }

    return text.slice(index + separator.length).trim()

}


