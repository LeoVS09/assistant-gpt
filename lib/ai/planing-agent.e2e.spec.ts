import {jest, describe, it, expect} from '@jest/globals';
import { DateTime } from 'luxon';

import { PlaningAgent } from "./planing-agent";
import { ConversationManager } from './conversation-manager';
import { generateAiAgent } from './ai-agent';
import { QueryCommand, SaveCommandArgs, UpdateCommandArgs } from './assistant-models';

const SECONDS = 1000;
jest.setTimeout(70 * SECONDS)


describe('PlaningAgent', () => {
  const now = DateTime.fromJSDate(new Date('Tue Apr 04 2023'))
  console.log('assume for tests current date is', now.toLocaleString())

  const ai = generateAiAgent()
  const conversation = new ConversationManager(ai)

  const agent = new PlaningAgent(conversation, now);

  it(`should answer save command on 'Need read book "Domain-Driven Design: Tackling Complexity in the Heart of Software" this month`, async () => {
    const text = 'Need read book "Domain-Driven Design: Tackling Complexity in the Heart of Software" this month'
    
    const {original, commands, response} = await agent.plan({text: text})

    expect([
      {'system': agent.systemMessage},
      {'user': text},
      {'assistant': original.text}
    ]).toMatchSnapshot();

    expect(commands).toHaveLength(1)
    expect(response).toBeDefined()

    const [command] = commands!
    expect(command.id).toBe('save')

    const args = command.args as SaveCommandArgs
    expect(args.text).toBeDefined()
    expect(args.entities).toContain('Domain-Driven Design: Tackling Complexity in the Heart of Software')
    expect(args.tags?.join(' ')).toContain('read')
    expect(args.status).toContain('to-do')
    expect(args.deadline).toBe(now.endOf('month').toLocaleString(DateTime.DATE_FULL))
  });

  it(`should answer save command on 'Need work on taxes this week'`, async () => {
    const text = 'Need work on taxes this week'

    const {original, commands, response} = await agent.plan({text: text})

    expect([
      {'system': agent.systemMessage},
      {'user': text},
      {'assistant': original.text}
    ]).toMatchSnapshot();

    expect(commands).toHaveLength(1)
    expect(response).toBeDefined()

    const [command] = commands!
    expect(command.id).toBe('save')

    const args = command.args as SaveCommandArgs
    expect(args.text).toBeDefined()
    expect(args.entities).toContain('taxes')
    expect(args.tags?.join(' ')).toContain('work')
    expect(args.status).toContain('to-do')
    expect(args.deadline).toBe(now.endOf('week').toLocaleString(DateTime.DATE_FULL))
  });

  it(`should answer save command on 'Why the sun looks yellow?'`, async () => {
    const text = 'Why the sun looks yellow?'
    
    const {original, commands, response} = await agent.plan({text: text})

    expect([
      {'system': agent.systemMessage},
      {'user': text},
      {'assistant': original.text}
    ]).toMatchSnapshot();

    expect(commands || []).toHaveLength(0)
    expect(response).toBeDefined()

  });

  it(`should answer query command on "What to read this month?"`, async () => {
    const text = 'What to read this month?'

    const {original, commands, response} = await agent.plan({text: text})

    expect([
      {'system': agent.systemMessage},
      {'user': text},
      {'assistant': original.text}
    ]).toMatchSnapshot();

    expect(commands).toHaveLength(1)
    expect(response).toBeDefined()

    const [command] = commands!
    expect(command.id).toBe('query')

    const args = command.args as QueryCommand['args']
    expect(args.status).toContain('to-do')
    expect(args.deadline).toEqual({ '$lte': now.endOf('month').toLocaleString(DateTime.DATE_FULL) })
  });

  it(`should answer update command on "I'm finished with taxes"`, async () => {
    const text = "I'm finished with taxes"

    const {original, commands, response} = await agent.plan({text: text})

    expect([
      {'system': agent.systemMessage},
      {'user': text},
      {'assistant': original.text}
    ]).toMatchSnapshot();

    expect(commands).toHaveLength(1)
    expect(response).toBeDefined()

    const [command] = commands!
    expect(command.id).toBe('update')

    const args = command.args as UpdateCommandArgs
    expect(args.filter).toBeDefined()
    expect(args.filter.entities).toBe('taxes')
    expect(args.filter.status).toBe('to-do')
    expect(args.update).toBeDefined()
    expect(args.update.$set.status).toBe('done')
  });
    
});