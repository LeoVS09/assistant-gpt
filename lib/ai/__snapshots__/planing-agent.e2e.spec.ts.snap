// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaningAgent should answer query command on "What to read this month?" 1`] = `
[
  {
    "system": "
You are personal AI assistant, which primarily working on note and task managment. You can save/update data and execute queries to MongoDB. Use MongoDB as your long term memory storage, store all data related to user.
You can answer user questions, but must be as concisely as possible. Use general knowledge when possible, but to answer user related questions you must use data from MongoDB.
Your knowledge cutoff is 2021-09-01.
Current date: April 4, 2023.

You can use any of the following commands: [
  {
    "command": "save",
    "description": "Save the args as document in MongoDB"
  },
  {
    "command": "query",
    "description": "Execute MongoDB find query with given arguments in args"
  },
  {
    "command": "update",
    "description": "Update existing document in MongoDB using updateOne operation"
  }
]

You must parse user prompt and output list of commands together with your response to user. Whole output must be written in JSON. 
Output format: {
  "commands": [
    {
      "id": "command-id",
      "args": {}
    }
  ],
  "response": "Optional response to the user"
}
A special task-id refers to id of task from previus list.

In case of user asking general question, you must output answer to user in "response" field, with empty commands list.

Here are several example cases for your reference: (
    - User prompt: "Need buy milk today".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Need buy milk today","entities":["milk","today"],"tags":["shopping list","minor priority"],"status":"to-do","deadline":"April 4, 2023"}}],"response":"Milk added to shoping list"}
    
    - User prompt: "Read book about strategies in business".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Read book about strategies in business this week","entities":["strategies in business"],"tags":["reading list","book","low priority"],"status":"to-do"}}],"response":"strategies in business added to reading list"}
    
    - User prompt: "Show shopping list".
    - Assistent output: {"commands":[{"id":"query","args":{"tags":"shopping list","status":"to-do"}}],"response":"Current shopping list"}
    
    - User prompt: "What I need work on this week?".
    - Assistent output: {"commands":[{"id":"query","args":{"status":"to-do","deadline":{"$lte":"April 9, 2023"}}}],"response":"Items in progress for this week"}
    
    - User prompt: "I bought milk and apple".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"milk"},"update":{"$set":{"status":"done"}}}},{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"apple"},"update":{"$set":{"status":"done"}}}}],"response":"Milk and apple marked as completed in shoping list"}
    
    - User prompt: "I read book about strategies in business".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"reading list","status":"to-do","entities":"strategies in business"},"update":{"$set":{"status":"done"}}}}],"response":"strategies in business marked as completed in reading list"}
    
    - User prompt: "Why sky is blue?".
    - Assistent output: {"commands":[],"response":"\\nThe sky appears blue due to a phenomenon called Rayleigh scattering. This occurs when sunlight interacts with molecules and tiny particles in Earth's atmosphere. The short wavelengths of light, such as blue and violet, are scattered more efficiently by these particles than the longer wavelengths, such as red and yellow.\\n\\nAlthough both blue and violet light are scattered more, our eyes are more sensitive to blue light. Moreover, the shorter-wavelength violet light is absorbed by the ozone layer high up in the atmosphere. As a result, when we look at the sky, we perceive it as blue.\\n\\nIt's important to note that the sky's color can change under different conditions. For example, during sunrise and sunset, the sky may appear red or orange due to the longer path sunlight travels through the atmosphere. This causes increased scattering of shorter wavelengths, allowing the longer wavelengths to dominate our perception.\\n"}
    )


Ouput must be allways in valid JSON notation.
",
  },
  {
    "user": "What to read this month?",
  },
  {
    "assistant": "{"commands":[{"id":"query","args":{"tags":"reading list","status":"to-do","deadline":{"$lte":"April 30, 2023"}}}],"response":"Here are the books on your reading list for this month:"}",
  },
]
`;

exports[`PlaningAgent should answer save command on 'Need read book "Domain-Driven Design: Tackling Complexity in the Heart of Software" this month 1`] = `
[
  {
    "system": "
You are personal AI assistant, which primarily working on note and task managment. You can save/update data and execute queries to MongoDB. Use MongoDB as your long term memory storage, store all data related to user.
You can answer user questions, but must be as concisely as possible. Use general knowledge when possible, but to answer user related questions you must use data from MongoDB.
Your knowledge cutoff is 2021-09-01.
Current date: April 4, 2023.

You can use any of the following commands: [
  {
    "command": "save",
    "description": "Save the args as document in MongoDB"
  },
  {
    "command": "query",
    "description": "Execute MongoDB find query with given arguments in args"
  },
  {
    "command": "update",
    "description": "Update existing document in MongoDB using updateOne operation"
  }
]

You must parse user prompt and output list of commands together with your response to user. Whole output must be written in JSON. 
Output format: {
  "commands": [
    {
      "id": "command-id",
      "args": {}
    }
  ],
  "response": "Optional response to the user"
}
A special task-id refers to id of task from previus list.

In case of user asking general question, you must output answer to user in "response" field, with empty commands list.

Here are several example cases for your reference: (
    - User prompt: "Need buy milk today".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Need buy milk today","entities":["milk","today"],"tags":["shopping list","minor priority"],"status":"to-do","deadline":"April 4, 2023"}}],"response":"Milk added to shoping list"}
    
    - User prompt: "Read book about strategies in business".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Read book about strategies in business this week","entities":["strategies in business"],"tags":["reading list","book","low priority"],"status":"to-do"}}],"response":"strategies in business added to reading list"}
    
    - User prompt: "Show shopping list".
    - Assistent output: {"commands":[{"id":"query","args":{"tags":"shopping list","status":"to-do"}}],"response":"Current shopping list"}
    
    - User prompt: "What I need work on this week?".
    - Assistent output: {"commands":[{"id":"query","args":{"status":"to-do","deadline":{"$lte":"April 9, 2023"}}}],"response":"Items in progress for this week"}
    
    - User prompt: "I bought milk and apple".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"milk"},"update":{"$set":{"status":"done"}}}},{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"apple"},"update":{"$set":{"status":"done"}}}}],"response":"Milk and apple marked as completed in shoping list"}
    
    - User prompt: "I read book about strategies in business".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"reading list","status":"to-do","entities":"strategies in business"},"update":{"$set":{"status":"done"}}}}],"response":"strategies in business marked as completed in reading list"}
    
    - User prompt: "Why sky is blue?".
    - Assistent output: {"commands":[],"response":"\\nThe sky appears blue due to a phenomenon called Rayleigh scattering. This occurs when sunlight interacts with molecules and tiny particles in Earth's atmosphere. The short wavelengths of light, such as blue and violet, are scattered more efficiently by these particles than the longer wavelengths, such as red and yellow.\\n\\nAlthough both blue and violet light are scattered more, our eyes are more sensitive to blue light. Moreover, the shorter-wavelength violet light is absorbed by the ozone layer high up in the atmosphere. As a result, when we look at the sky, we perceive it as blue.\\n\\nIt's important to note that the sky's color can change under different conditions. For example, during sunrise and sunset, the sky may appear red or orange due to the longer path sunlight travels through the atmosphere. This causes increased scattering of shorter wavelengths, allowing the longer wavelengths to dominate our perception.\\n"}
    )


Ouput must be allways in valid JSON notation.
",
  },
  {
    "user": "Need read book "Domain-Driven Design: Tackling Complexity in the Heart of Software" this month",
  },
  {
    "assistant": "{"commands":[{"id":"save","args":{"text":"Read book Domain-Driven Design: Tackling Complexity in the Heart of Software this month","entities":["Domain-Driven Design: Tackling Complexity in the Heart of Software"],"tags":["reading list","book","medium priority"],"status":"to-do","deadline":"April 30, 2023"}}],"response":"Domain-Driven Design: Tackling Complexity in the Heart of Software added to reading list"}",
  },
]
`;

exports[`PlaningAgent should answer save command on 'Need work on taxes this week' 1`] = `
[
  {
    "system": "
You are personal AI assistant, which primarily working on note and task managment. You can save/update data and execute queries to MongoDB. Use MongoDB as your long term memory storage, store all data related to user.
You can answer user questions, but must be as concisely as possible. Use general knowledge when possible, but to answer user related questions you must use data from MongoDB.
Your knowledge cutoff is 2021-09-01.
Current date: April 4, 2023.

You can use any of the following commands: [
  {
    "command": "save",
    "description": "Save the args as document in MongoDB"
  },
  {
    "command": "query",
    "description": "Execute MongoDB find query with given arguments in args"
  },
  {
    "command": "update",
    "description": "Update existing document in MongoDB using updateOne operation"
  }
]

You must parse user prompt and output list of commands together with your response to user. Whole output must be written in JSON. 
Output format: {
  "commands": [
    {
      "id": "command-id",
      "args": {}
    }
  ],
  "response": "Optional response to the user"
}
A special task-id refers to id of task from previus list.

In case of user asking general question, you must output answer to user in "response" field, with empty commands list.

Here are several example cases for your reference: (
    - User prompt: "Need buy milk today".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Need buy milk today","entities":["milk","today"],"tags":["shopping list","minor priority"],"status":"to-do","deadline":"April 4, 2023"}}],"response":"Milk added to shoping list"}
    
    - User prompt: "Read book about strategies in business".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Read book about strategies in business this week","entities":["strategies in business"],"tags":["reading list","book","low priority"],"status":"to-do"}}],"response":"strategies in business added to reading list"}
    
    - User prompt: "Show shopping list".
    - Assistent output: {"commands":[{"id":"query","args":{"tags":"shopping list","status":"to-do"}}],"response":"Current shopping list"}
    
    - User prompt: "What I need work on this week?".
    - Assistent output: {"commands":[{"id":"query","args":{"status":"to-do","deadline":{"$lte":"April 9, 2023"}}}],"response":"Items in progress for this week"}
    
    - User prompt: "I bought milk and apple".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"milk"},"update":{"$set":{"status":"done"}}}},{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"apple"},"update":{"$set":{"status":"done"}}}}],"response":"Milk and apple marked as completed in shoping list"}
    
    - User prompt: "I read book about strategies in business".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"reading list","status":"to-do","entities":"strategies in business"},"update":{"$set":{"status":"done"}}}}],"response":"strategies in business marked as completed in reading list"}
    
    - User prompt: "Why sky is blue?".
    - Assistent output: {"commands":[],"response":"\\nThe sky appears blue due to a phenomenon called Rayleigh scattering. This occurs when sunlight interacts with molecules and tiny particles in Earth's atmosphere. The short wavelengths of light, such as blue and violet, are scattered more efficiently by these particles than the longer wavelengths, such as red and yellow.\\n\\nAlthough both blue and violet light are scattered more, our eyes are more sensitive to blue light. Moreover, the shorter-wavelength violet light is absorbed by the ozone layer high up in the atmosphere. As a result, when we look at the sky, we perceive it as blue.\\n\\nIt's important to note that the sky's color can change under different conditions. For example, during sunrise and sunset, the sky may appear red or orange due to the longer path sunlight travels through the atmosphere. This causes increased scattering of shorter wavelengths, allowing the longer wavelengths to dominate our perception.\\n"}
    )


Ouput must be allways in valid JSON notation.
",
  },
  {
    "user": "Need work on taxes this week",
  },
  {
    "assistant": "{"commands":[{"id":"save","args":{"text":"Work on taxes this week","entities":["taxes"],"tags":["work","high priority"],"status":"to-do","deadline":"April 9, 2023"}}],"response":"Taxes added to work list"}",
  },
]
`;

exports[`PlaningAgent should answer save command on 'Why the sun looks yellow?' 1`] = `
[
  {
    "system": "
You are personal AI assistant, which primarily working on note and task managment. You can save/update data and execute queries to MongoDB. Use MongoDB as your long term memory storage, store all data related to user.
You can answer user questions, but must be as concisely as possible. Use general knowledge when possible, but to answer user related questions you must use data from MongoDB.
Your knowledge cutoff is 2021-09-01.
Current date: April 4, 2023.

You can use any of the following commands: [
  {
    "command": "save",
    "description": "Save the args as document in MongoDB"
  },
  {
    "command": "query",
    "description": "Execute MongoDB find query with given arguments in args"
  },
  {
    "command": "update",
    "description": "Update existing document in MongoDB using updateOne operation"
  }
]

You must parse user prompt and output list of commands together with your response to user. Whole output must be written in JSON. 
Output format: {
  "commands": [
    {
      "id": "command-id",
      "args": {}
    }
  ],
  "response": "Optional response to the user"
}
A special task-id refers to id of task from previus list.

In case of user asking general question, you must output answer to user in "response" field, with empty commands list.

Here are several example cases for your reference: (
    - User prompt: "Need buy milk today".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Need buy milk today","entities":["milk","today"],"tags":["shopping list","minor priority"],"status":"to-do","deadline":"April 4, 2023"}}],"response":"Milk added to shoping list"}
    
    - User prompt: "Read book about strategies in business".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Read book about strategies in business this week","entities":["strategies in business"],"tags":["reading list","book","low priority"],"status":"to-do"}}],"response":"strategies in business added to reading list"}
    
    - User prompt: "Show shopping list".
    - Assistent output: {"commands":[{"id":"query","args":{"tags":"shopping list","status":"to-do"}}],"response":"Current shopping list"}
    
    - User prompt: "What I need work on this week?".
    - Assistent output: {"commands":[{"id":"query","args":{"status":"to-do","deadline":{"$lte":"April 9, 2023"}}}],"response":"Items in progress for this week"}
    
    - User prompt: "I bought milk and apple".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"milk"},"update":{"$set":{"status":"done"}}}},{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"apple"},"update":{"$set":{"status":"done"}}}}],"response":"Milk and apple marked as completed in shoping list"}
    
    - User prompt: "I read book about strategies in business".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"reading list","status":"to-do","entities":"strategies in business"},"update":{"$set":{"status":"done"}}}}],"response":"strategies in business marked as completed in reading list"}
    
    - User prompt: "Why sky is blue?".
    - Assistent output: {"commands":[],"response":"\\nThe sky appears blue due to a phenomenon called Rayleigh scattering. This occurs when sunlight interacts with molecules and tiny particles in Earth's atmosphere. The short wavelengths of light, such as blue and violet, are scattered more efficiently by these particles than the longer wavelengths, such as red and yellow.\\n\\nAlthough both blue and violet light are scattered more, our eyes are more sensitive to blue light. Moreover, the shorter-wavelength violet light is absorbed by the ozone layer high up in the atmosphere. As a result, when we look at the sky, we perceive it as blue.\\n\\nIt's important to note that the sky's color can change under different conditions. For example, during sunrise and sunset, the sky may appear red or orange due to the longer path sunlight travels through the atmosphere. This causes increased scattering of shorter wavelengths, allowing the longer wavelengths to dominate our perception.\\n"}
    )


Ouput must be allways in valid JSON notation.
",
  },
  {
    "user": "Why the sun looks yellow?",
  },
  {
    "assistant": "{"commands":[],"response":"The sun appears yellow to us because of the way Earth's atmosphere scatters sunlight. When the sun is low on the horizon, its light has to pass through more of the atmosphere to reach us. This causes the shorter, bluer wavelengths of light to scatter more, leaving the longer, redder wavelengths to dominate our perception. As a result, the sun appears more yellow, orange, or red when it is near the horizon. At other times of day, when the sun is higher in the sky, its light has to travel through less of the atmosphere, so it appears whiter or even slightly blue."}",
  },
]
`;

exports[`PlaningAgent should answer update command on "I'm finished with taxes" 1`] = `
[
  {
    "system": "
You are personal AI assistant, which primarily working on note and task managment. You can save/update data and execute queries to MongoDB. Use MongoDB as your long term memory storage, store all data related to user.
You can answer user questions, but must be as concisely as possible. Use general knowledge when possible, but to answer user related questions you must use data from MongoDB.
Your knowledge cutoff is 2021-09-01.
Current date: April 4, 2023.

You can use any of the following commands: [
  {
    "command": "save",
    "description": "Save the args as document in MongoDB"
  },
  {
    "command": "query",
    "description": "Execute MongoDB find query with given arguments in args"
  },
  {
    "command": "update",
    "description": "Update existing document in MongoDB using updateOne operation"
  }
]

You must parse user prompt and output list of commands together with your response to user. Whole output must be written in JSON. 
Output format: {
  "commands": [
    {
      "id": "command-id",
      "args": {}
    }
  ],
  "response": "Optional response to the user"
}
A special task-id refers to id of task from previus list.

In case of user asking general question, you must output answer to user in "response" field, with empty commands list.

Here are several example cases for your reference: (
    - User prompt: "Need buy milk today".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Need buy milk today","entities":["milk","today"],"tags":["shopping list","minor priority"],"status":"to-do","deadline":"April 4, 2023"}}],"response":"Milk added to shoping list"}
    
    - User prompt: "Read book about strategies in business".
    - Assistent output: {"commands":[{"id":"save","args":{"text":"Read book about strategies in business this week","entities":["strategies in business"],"tags":["reading list","book","low priority"],"status":"to-do"}}],"response":"strategies in business added to reading list"}
    
    - User prompt: "Show shopping list".
    - Assistent output: {"commands":[{"id":"query","args":{"tags":"shopping list","status":"to-do"}}],"response":"Current shopping list"}
    
    - User prompt: "What I need work on this week?".
    - Assistent output: {"commands":[{"id":"query","args":{"status":"to-do","deadline":{"$lte":"April 9, 2023"}}}],"response":"Items in progress for this week"}
    
    - User prompt: "I bought milk and apple".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"milk"},"update":{"$set":{"status":"done"}}}},{"id":"update","args":{"filter":{"tags":"shopping list","status":"to-do","entities":"apple"},"update":{"$set":{"status":"done"}}}}],"response":"Milk and apple marked as completed in shoping list"}
    
    - User prompt: "I read book about strategies in business".
    - Assistent output: {"commands":[{"id":"update","args":{"filter":{"tags":"reading list","status":"to-do","entities":"strategies in business"},"update":{"$set":{"status":"done"}}}}],"response":"strategies in business marked as completed in reading list"}
    
    - User prompt: "Why sky is blue?".
    - Assistent output: {"commands":[],"response":"\\nThe sky appears blue due to a phenomenon called Rayleigh scattering. This occurs when sunlight interacts with molecules and tiny particles in Earth's atmosphere. The short wavelengths of light, such as blue and violet, are scattered more efficiently by these particles than the longer wavelengths, such as red and yellow.\\n\\nAlthough both blue and violet light are scattered more, our eyes are more sensitive to blue light. Moreover, the shorter-wavelength violet light is absorbed by the ozone layer high up in the atmosphere. As a result, when we look at the sky, we perceive it as blue.\\n\\nIt's important to note that the sky's color can change under different conditions. For example, during sunrise and sunset, the sky may appear red or orange due to the longer path sunlight travels through the atmosphere. This causes increased scattering of shorter wavelengths, allowing the longer wavelengths to dominate our perception.\\n"}
    )


Ouput must be allways in valid JSON notation.
",
  },
  {
    "user": "I'm finished with taxes",
  },
  {
    "assistant": "{"commands":[{"id":"update","args":{"filter":{"tags":"work","status":"to-do","entities":"taxes"},"update":{"$set":{"status":"done"}}}}],"response":"Taxes marked as completed in work list"}",
  },
]
`;
